### PR TITLE
Add NetApp SolidFire volume module

### PR DESCRIPTION
**Pull request by timuster**
##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

NetApp SolidFire volume manager module

##### ANSIBLE VERSION

```
ansible 2.3.0
  config file = 
  configured module search path = Default w/o overrides
```

##### SUMMARY

Adding sf_volume_manager.py, a module to manage NetApp SolidFire volumes. The design was based on the existing NetApp E-Series modules.

```

```

### CODE REVIEW

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (hulquest) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Since it's a new file, I think the copyright should be 2017.

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (hulquest) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Strike

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (hulquest) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Is this necessary given account_id is a required parameter?  My guidance is to eliminate this parameter and clearly comment why volume names can be used.  
Along the same lines, it's always s fine line between doing what's easiest in the api and whats easier for the playbook developer.   It's not a must-fix for this PR, but I would like to understand the impact to this module (probably all SF modules) and the associated playbooks of allowing account names to be used instead of account id's.  Are the playbooks easier to read or develop?  Or does this module nicety amount to needless lines of code in the module?  Getting a module out in to the community now is more important than making this decision.

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (hulquest) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Is DEBUG the appropriate level?

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (hulquest) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: I believe the previous 3 lines can be deleted.

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (hulquest) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Call out default?  If you set a default here, don't forget to update the spec above.

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (hulquest) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Should the volume be deleted here?

Comment (hulquest) [reply lib/ansible/modules/storage/netapp/sf_volume_manager.py]: I see the pattern.  Disregard volume delete comment.

**Review by hulquest — CHANGES_REQUESTED** [review]
I made some comments that should require minor fixes.  

Personally, I don't like the pattern of determining if the object is changed and then acting upon it in a subsequent block of code.  I don't think that should hold up getting this module into the community though.

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (timuster) [reply lib/ansible/modules/storage/netapp/sf_volume_manager.py]: I'll get this fixed across all PRs. Thanks for pointing it out!

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (timuster) [reply lib/ansible/modules/storage/netapp/sf_volume_manager.py]: This is to keep it consistent with the existing E-Series modules.

**Comment by timuster** [comment]
@hulquest Thank you for the review! I chose to go with the code pattern of determining a change and then acting upon it in a subsequent block of code since 1) It allows us to support check_mode (and to avoid making any changes if check_mode is enabled), and 2) to keep things consistent with the already published NetApp E-Series modules. Here's an example: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/storage/netapp/netapp_e_volume.py

I'll fix the other points you brought up as soon as possible. Thank you once again!

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (timuster) [reply lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Explaining the design choice in my comment below.

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (timuster) [reply lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Volume names can be used. The reason I added this optional volume_id parameter was to allow an end-user to distinguish between 2 volumes with the same name, but different volume IDs - something that the SolidFire manageability GUI allows. I like the suggestion regarding using account name instead of account IDs, but I think we'll again need both parameters for the same reason as above (same name, different IDs)... Please let me know if my assumption is incorrect. However, like you mentioned, we can make these improvements after this PR has merged.

**Comment by hulquest** [comment]
LGTM from a SF point of view.  shipit

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (resmo) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: could you but this in the description as a new item like 
~~~yaml
        description:
        - Should the volume provides 512-byte sector emulation?
        - Required when C(state=present)
~~~

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (resmo) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: in many modules the default is `state=present` please consider it too if it makes sense.

**Comment by timuster** [comment]
@resmo Thank you for your review! I have made the requested changes.

**Comment by timuster** [comment]
ready_for_review

**Comment by hulquest** [comment]
bot_status

**Comment by hulquest** [comment]
shipit 
Want to move the review along.

**Comment by timuster** [comment]
bot_status

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (resmo) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: it looks to me this module supports check mode, but it is says not

**Comment by timuster** [comment]
ready_for_review

**Comment by resmo** [comment]
shipit

**Comment by hulquest** [comment]
shipit

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (bcoca) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: new_X is a bad pattern, if you have a unique identifier a comparison of account_id supplied vs existing should determine if this changes or not

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (bcoca) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: enable is imperative 512emulation: yes|no is a more declarative

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (bcoca) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: logging and debug framework already exists for modules, it also takes care of many other options that affect this, do not roll out your own

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (bcoca) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: as explained in related ticket, catchall exceptions are not the best way to handle user errors

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (bcoca) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: a module should never end in exception, it should use fail_json instead (you can pass traceback info if needed in 'exception' key)

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: No need to specify `type:` in documentation, we don't use that value.

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: All descriptions should be full sentences, e.g start with a capital letter and end with a full stop.

**Comment by timuster** [comment]
ready_for_review

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Although not required, we generally put `description:` first as we find it makes the source code a little more readable.

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: No need to document `type:`, you can remove these lines

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (gundalow) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: `I(volume_id)`

File: lib/ansible/utils/module_docs_fragments/netapp.py
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/netapp.py]: I see you have this change in a number of PRs. Once the first PR gets merged and you rebase the remaining PRs this change should disappear from the remaining PRs.

**Review by gundalow — CHANGES_REQUESTED** [review]

File: lib/ansible/utils/module_docs_fragments/netapp.py
Comment (timuster) [reply lib/ansible/utils/module_docs_fragments/netapp.py]: Yeah, I'll rebase the other PRs once one of them gets merged.

**Review by gundalow — APPROVED** [review]

File: lib/ansible/utils/module_docs_fragments/netapp.py
Comment (gundalow) [reply lib/ansible/utils/module_docs_fragments/netapp.py]: I've merged #20734 so you can go ahead and rebase this now

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (resmo) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py:54]: I didn't see such a argument in arg_spec, is this `enable512e`? Note you can make aliases in arg_spec if you like to keep the logic. See my comment in the arg spec. Add the arg_spec variable name as alias here:
~~~yaml
512emulation: 
    description:  
    - Should the volume provide 512-byte sector emulation?
    - Required when C(state=present)
    required: false
    aliases: ['enable512e']
~~~

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (resmo) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: Docs uses `512emulation`: add an alias if you like to keep the code as is 
~~~python
enable512e=dict(type='bool', aliases=['512emulation']),
~~~

**Review by resmo — CHANGES_REQUESTED** [review]

**Review by resmo — APPROVED** [review]

**Comment by resmo** [comment]
shipit

File: lib/ansible/modules/storage/netapp/sf_volume_manager.py
Comment (bcoca) [inline lib/ansible/modules/storage/netapp/sf_volume_manager.py]: this will just confuse people and they'll try to provide a json string, just change to 'takes a yaml dictionary', also, not a list as you have` type='dict' `in arg spec

**Comment by timuster** [comment]
ready_for_review

**Review by gundalow — APPROVED** [review]

**Pull request merged**
